File: apps/api/app/schemas/synastry.py
# ############################################################################
# AI_HEADER: MODULE_SCHEMAS_SYNASTRY
# ROLE: Pydantic schemas for synastry (compatibility) endpoints
# DEPENDENCIES: pydantic, app.schemas._base.CamelModel
# GRACE_ANCHORS: []
# ############################################################################

# START_MODULE_CONTRACT: M-SCHEMAS-SYNASTRY
# purpose: Pydantic schemas for synastry capabilities, partners, reports, aspect details, and feedback.
# owns:
#   - apps/api/app/schemas/synastry.py
# inputs: request payload dicts / query params
# outputs: validated models in camelCase wire format
# dependencies: app.schemas._base.CamelModel
# side_effects: none (pure schemas)
# emitted_logs: none
# invariants:
#   - All public schemas subclass CamelModel for camelCase wire format
# failure_policy: Pydantic ValidationError on invalid data
# END_MODULE_CONTRACT: M-SCHEMAS-SYNASTRY

# START_MODULE_MAP: M-SCHEMAS-SYNASTRY
# public_entrypoints:
#   - SynastryCapabilitiesRead
#   - PartnerCreate
#   - SynastryPartnerItem
#   - SynastryListRead
#   - SynastryGenerationRead
#   - SynastryAspect
#   - AspectDrilldown
#   - SynastrySphere
#   - SynastryReport
#   - SynastryFeedbackWrite
#   - SynastryFeedbackRead
# semantic_blocks: none
# owned_tests: none
# END_MODULE_MAP: M-SCHEMAS-SYNASTRY

from __future__ import annotations

from datetime import date, datetime, time
from decimal import Decimal
import uuid
from typing import Any, Literal

from pydantic import Field

from app.schemas._base import CamelModel


class SynastryCapabilitiesRead(CamelModel):
    """Capabilities status for synastry feature."""

    can_calculate: bool = Field(..., description="Whether user can calculate new synastry report")
    active_partner_count: int = Field(..., description="Current count of active partners")
    max_partners: int = Field(default=20, description="Maximum allowed active partners per user")
    has_unlocked_access: bool = Field(..., description="Whether user has unlocked synastry access")
    credit_balance: int = Field(default=0, description="Available credit balance")


class PartnerCreate(CamelModel):
    """Request payload to create a new synastry partner."""

    name: str = Field(..., min_length=1, max_length=120, description="Partner name or nickname")
    relation: str = Field(default="romantic", description="Relation type (romantic, friend, business, family)")
    birth_date: date = Field(..., description="Birth date (YYYY-MM-DD)")
    birth_time: time | None = Field(default=None, description="Birth time (HH:MM:SS or HH:MM)")
    birth_city: str | None = Field(default=None, description="Birth city name")
    birth_lat: Decimal | float | None = Field(default=None, ge=-90, le=90, description="Birth latitude")
    birth_lon: Decimal | float | None = Field(default=None, ge=-180, le=180, description="Birth longitude")
    birth_tz: str | None = Field(default=None, description="Birth IANA timezone string")
    birth_time_precision: Literal["exact", "approximate"] = Field(
        default="exact", description="Birth time precision: exact or approximate"
    )
    idempotency_key: str | None = Field(default=None, max_length=80, description="Idempotency key")


class SynastryPartnerItem(CamelModel):
    """Summary item for a partner in the synastry list."""

    id: uuid.UUID = Field(..., description="Partner UUID")
    name: str = Field(..., description="Partner name")
    relation_type: str = Field(..., description="Relation type")
    birth_date: date = Field(..., description="Birth date")
    precision: str = Field(..., description="Birth time precision (exact/approximate)")
    score: int | None = Field(default=None, description="Overall compatibility score (0..100)")
    status: Literal["good", "mid", "bad"] | None = Field(default=None, description="Valence status")
    summary: str | None = Field(default=None, description="Short summary verdict")
    counters: dict[str, int] | None = Field(default=None, description="Counts of good, mid, bad aspects")
    report_state: str | None = Field(default=None, description="Report calculation state (pending, calculating, ready, failed)")
    created_at: datetime = Field(..., description="Partner creation timestamp")


class SynastryListRead(CamelModel):
    """List response of user's synastry partners."""

    partners: list[SynastryPartnerItem] = Field(default_factory=list, description="List of partner items")


class SynastryGenerationRead(CamelModel):
    """Status read model for async report calculation generation."""

    report_id: uuid.UUID = Field(..., description="Report UUID")
    partner_id: uuid.UUID = Field(..., description="Partner UUID")
    state: Literal["pending", "calculating", "narrative_generating", "ready", "failed", "invalidated"] = Field(
        ..., description="Current calculation state"
    )
    stage: str | None = Field(default=None, description="Human-readable processing stage")
    attempt_count: int = Field(default=0, description="Current calculation attempt count")
    error_code: str | None = Field(default=None, description="Error code if state is failed")
    error_message: str | None = Field(default=None, description="Public error message if failed")


class SynastryAspect(CamelModel):
    """Synastry aspect entry for interactive wheel and list."""

    id: str = Field(..., description="Aspect ID e.g. sun_trine_moon")
    title: str = Field(..., description="Human aspect title e.g. Солнце трин Луна")
    tone: Literal["good", "mid", "bad", "harmony", "tension", "neutral", "supportive", "mixed", "tense"] = Field(..., description="Aspect tone/valence")
    score: int | None = Field(default=None, description="Aspect impact score")
    description: str | None = Field(default=None, description="Short interpretation snippet")
    tech_signature: str | None = Field(default=None, description="Technical aspect signature")


class AspectDrilldown(CamelModel):
    """Detailed drill-down interpretation for a single synastry aspect."""

    aspect_id: str = Field(..., description="Aspect ID")
    title: str = Field(..., description="Aspect title")
    tone: Literal["good", "mid", "bad", "harmony", "tension", "neutral", "supportive", "mixed", "tense"] = Field(..., description="Aspect tone")
    tech_signature: str | None = Field(default=None, description="Technical astrological signature")
    explanation: str = Field(..., description="Deep psychological and dynamic explanation")
    scenario: str | None = Field(default=None, description="Real-life interaction scenario")
    advice: str | None = Field(default=None, description="Constructive relationship advice")


class SynastrySphere(CamelModel):
    """Synastry breakdown sphere (Intimacy, Communication, Daily Life, Work & Money)."""

    id: str = Field(..., description="Sphere identifier e.g. intimacy, communication, daily_life, finance")
    title: str = Field(..., description="Display title e.g. Близость")
    score: int = Field(..., ge=0, le=100, description="Sphere compatibility score (0..100)")
    description: str | None = Field(default=None, description="Detailed sphere explanation")


class SynastryReport(CamelModel):
    """Full synastry report data model."""

    id: uuid.UUID = Field(..., description="Report UUID")
    owner_id: uuid.UUID = Field(..., description="Owner user UUID")
    partner_id: uuid.UUID = Field(..., description="Partner UUID")
    partner_name: str = Field(..., description="Partner name")
    relation_type: str = Field(..., description="Relation type")
    precision: str = Field(..., description="Birth time precision (exact/approximate)")
    score: int = Field(..., ge=0, le=100, description="Overall compatibility score (0..100)")
    status: Literal["good", "mid", "bad"] = Field(..., description="Overall compatibility valence status")
    verdict: str = Field(..., description="One-line verdict headline")
    summary: str = Field(..., description="Paragraph executive summary")
    hero_title: str | None = Field(default=None, description="Hero section headline")
    hero_description: str | None = Field(default=None, description="Hero section subtitle")
    counters: dict[str, int] = Field(default_factory=dict, description="Counts of good, mid, bad aspects")
    aspects: list[SynastryAspect] = Field(default_factory=list, description="Key synastry aspects list")
    house_overlays: list[dict[str, Any]] = Field(default_factory=list, description="House overlay interpretations")
    spheres: list[SynastrySphere] = Field(default_factory=list, description="Spheres breakdown")
    translations: list[dict[str, Any]] = Field(default_factory=list, description="Human translation cards")
    user_feedback: str | None = Field(default=None, description="Current user reality feedback value")
    created_at: datetime = Field(..., description="Report creation timestamp")


class SynastryFeedbackWrite(CamelModel):
    """Request payload to submit reality check feedback."""

    value: str = Field(..., min_length=1, max_length=30, description="Feedback choice (accurate, partial, inaccurate)")


class SynastryFeedbackRead(CamelModel):
    """Response payload for submitted feedback."""

    report_id: uuid.UUID = Field(..., description="Report UUID")
    value: str = Field(..., description="Feedback value")
    updated_at: datetime = Field(..., description="Submission timestamp")

File: lib/api/synastry.ts
// ############################################################################
// AI_HEADER: FRONTEND_API_SYNASTRY — typed synastry API client.
// ROLE: Typed facade for /api/synastry endpoints consumed by synastry UI.
// DEPENDENCIES: lib/log/instrumented-fetch
// ############################################################################

// START_MODULE_CONTRACT: M-FRONTEND-API-SYNASTRY
// purpose: Fetch capabilities, list partners, report details, aspect drilldowns, and submit feedback.
// owns:
//   - lib/api/synastry.ts
// inputs: Partner creation parameters, partner_id, aspect_id, feedback value.
// outputs: Synastry capabilities, partners list, generation status, report detail, aspect drilldown.
// dependencies: lib/log/instrumented-fetch.
// side_effects: credentialed synastry API HTTP requests.
// emitted_logs: none
// failure_policy: Throws SynastryApiError on non-2xx responses.
// END_MODULE_CONTRACT: M-FRONTEND-API-SYNASTRY

// START_MODULE_MAP: M-FRONTEND-API-SYNASTRY
// public_entrypoints:
//   - SynastryApiError
//   - getSynastryCapabilities
//   - getSynastryPartners
//   - createSynastryPartner
//   - getSynastryReport
//   - getSynastryStatus
//   - getAspectDrilldown
//   - submitSynastryFeedback
//   - deleteSynastryPartner
// semantic_blocks: none
// owned_tests: none
// END_MODULE_MAP: M-FRONTEND-API-SYNASTRY

import { instrumentedFetch } from "@/lib/log/instrumented-fetch"

export class SynastryApiError extends Error {
  status: number
  code: string | null

  constructor({ status, code, message }: { status: number; code?: string | null; message: string }) {
    super(message)
    this.name = "SynastryApiError"
    this.status = status
    this.code = code ?? null
  }
}

async function buildSynastryApiError(res: Response): Promise<SynastryApiError> {
  const status = res.status
  let code: string | null = null
  let message = `Request failed with status ${status}`

  try {
    const data = await res.clone().json()
    const detail = data?.detail
    if (detail) {
      if (typeof detail === "string") {
        message = detail
        code = detail
      } else if (typeof detail === "object" && detail !== null) {
        if (typeof detail.message === "string") message = detail.message
        if (typeof detail.code === "string") code = detail.code
      }
    } else if (data?.message && typeof data.message === "string") {
      message = data.message
      if (typeof data.code === "string") code = data.code
    }
  } catch {
    // Ignore JSON parse errors
  }

  return new SynastryApiError({ status, code, message })
}

export type SynastryCapabilities = {
  canCalculate: boolean
  canPurchase?: boolean
  activePartnerCount: number
  maxPartners: number
  hasUnlockedAccess: boolean
  creditBalance: number
  blockedReason?: "no_credits" | "partner_limit" | null
}

export type PartnerCreatePayload = {
  name: string
  relation?: string
  birthDate: string
  birthTime?: string | null
  birthCity?: string | null
  birthLat?: number | null
  birthLon?: number | null
  birthTz?: string | null
  birthTimePrecision?: "exact" | "approximate"
  idempotencyKey?: string | null
}

export type SynastryPartnerListItem = {
  id: string
  name: string
  relationType: string
  birthDate: string
  precision: "exact" | "approximate"
  score: number | null
  status: "good" | "mid" | "bad" | null
  summary: string | null
  counters?: { good: number; mid: number; bad: number } | null
  reportState?: string | null
  createdAt: string
}

export type SynastryGenerationStatus = {
  reportId: string
  partnerId: string
  state: "pending" | "calculating" | "narrative_generating" | "ready" | "failed" | "invalidated"
  stage: string | null
  attemptCount: number
  errorCode: string | null
  errorMessage: string | null
}

export type SynastryPlanetPoint = {
  id: string
  owner: "user" | "partner"
  planet: string
  longitude: number
  sign?: string | null
  retrograde?: boolean
  house?: number | null
  houseReliable?: boolean
}

export type SynastryAspectItem = {
  id: string
  title: string
  tone: "good" | "mid" | "bad" | "supportive" | "mixed" | "tense"
  score: number | null
  description: string | null
  techSignature: string | null
  ownerPlanetKey?: string | null
  partnerPlanetKey?: string | null
  aspectSymbol?: string | null
  orbDegrees?: number | null
  orbLabel?: string | null
}

export type SynastrySphereItem = {
  id: string
  title: string
  score: number
  description: string
}

export type SynastryTranslation = {
  aspectId?: string
  title: string
  tone: string
  tech?: string
  text?: string
  scene?: string
}

export type SynastryReportData = {
  partnerId: string
  partnerName: string
  relationType: string
  partnerBirthDate: string
  partnerBirthTime: string | null
  partnerBirthCity: string | null
  precision: "exact" | "approximate"
  score: number
  status: "good" | "mid" | "bad"
  verdict: string
  summary: string
  heroTitle: string | null
  heroDescription: string | null
  counters: { good: number; mid: number; bad: number }
  ownerPlanets: SynastryPlanetPoint[]
  partnerPlanets: SynastryPlanetPoint[]
  aspects: SynastryAspectItem[]
  houseOverlays: Array<{ tech?: string; text?: string }>
  houseSystem?: string
  translations: SynastryTranslation[]
  spheres: SynastrySphereItem[]
  userFeedback: string | null
  createdAt: string
}

export type AspectDrilldownData = {
  aspectId: string
  title: string
  tone: "good" | "mid" | "bad" | "supportive" | "mixed" | "tense"
  techSignature?: string
  aspectSymbol?: string
  aspectKindLabel?: string
  orbText?: string
  headline?: string
  ownerPlanet?: { key: string; label: string; glyph: string; meaning: string }
  partnerPlanet?: { key: string; label: string; glyph: string; meaning: string }
  aspectMechanics?: string
  explanation?: string
  scenes?: Array<{ title: string; text: string }>
  repairs?: string[]
  notMeans?: string[]
  scenario?: string
  advice?: string
}

export async function getSynastryCapabilities(): Promise<SynastryCapabilities> {
  const res = await instrumentedFetch({
    operation: "getSynastryCapabilities",
    routeTemplate: "/api/synastry/capabilities",
    url: "/api/synastry/capabilities",
    init: { credentials: "include" },
  })
  if (!res.ok) {
    throw await buildSynastryApiError(res)
  }
  return res.json()
}

export async function getSynastryPartners(): Promise<SynastryPartnerListItem[]> {
  const res = await instrumentedFetch({
    operation: "getSynastryPartners",
    routeTemplate: "/api/synastry",
    url: "/api/synastry",
    init: { credentials: "include" },
  })
  if (!res.ok) {
    throw await buildSynastryApiError(res)
  }
  const data = await res.json()
  return data.partners || []
}

export async function createSynastryPartner(
  payload: PartnerCreatePayload
): Promise<SynastryGenerationStatus> {
  const res = await instrumentedFetch({
    operation: "createSynastryPartner",
    routeTemplate: "/api/synastry/partners",
    url: "/api/synastry/partners",
    init: {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      credentials: "include",
      body: JSON.stringify(payload),
    },
  })
  if (!res.ok) {
    throw await buildSynastryApiError(res)
  }
  return res.json()
}

export async function getSynastryReport(partnerId: string): Promise<SynastryReportData> {
  const res = await instrumentedFetch({
    operation: "getSynastryReport",
    routeTemplate: "/api/synastry/{partner_id}",
    url: `/api/synastry/${partnerId}`,
    init: { credentials: "include" },
  })
  if (!res.ok) {
    throw await buildSynastryApiError(res)
  }
  return res.json()
}

export async function getSynastryStatus(partnerId: string): Promise<SynastryGenerationStatus> {
  const res = await instrumentedFetch({
    operation: "getSynastryStatus",
    routeTemplate: "/api/synastry/{partner_id}/status",
    url: `/api/synastry/${partnerId}/status`,
    init: { credentials: "include" },
  })
  if (!res.ok) {
    throw await buildSynastryApiError(res)
  }
  return res.json()
}

export async function getAspectDrilldown(
  partnerId: string,
  aspectId: string
): Promise<AspectDrilldownData> {
  const res = await instrumentedFetch({
    operation: "getAspectDrilldown",
    routeTemplate: "/api/synastry/{partner_id}/aspect/{aspect_id}",
    url: `/api/synastry/${partnerId}/aspect/${aspectId}`,
    init: { credentials: "include" },
  })
  if (!res.ok) {
    throw await buildSynastryApiError(res)
  }
  return res.json()
}

export async function submitSynastryFeedback(
  partnerId: string,
  value: string
): Promise<{ reportId: string; value: string; updatedAt: string }> {
  const res = await instrumentedFetch({
    operation: "submitSynastryFeedback",
    routeTemplate: "/api/synastry/{partner_id}/feedback",
    url: `/api/synastry/${partnerId}/feedback`,
    init: {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      credentials: "include",
      body: JSON.stringify({ value }),
    },
  })
  if (!res.ok) {
    throw await buildSynastryApiError(res)
  }
  return res.json()
}

export async function deleteSynastryPartner(partnerId: string): Promise<void> {
  const res = await instrumentedFetch({
    operation: "deleteSynastryPartner",
    routeTemplate: "/api/synastry/{partner_id}",
    url: `/api/synastry/${partnerId}`,
    init: {
      method: "DELETE",
      credentials: "include",
    },
  })
  if (!res.ok) {
    throw await buildSynastryApiError(res)
  }
}
